hpy/debug/src/memprotect.c

```diff
@@ -0,0 +1,60 @@
+#include "debug_internal.h"
+
+// Implements OS dependent abstraction of memory protection
+
+#ifdef _HPY_DEBUG_MEM_PROTECT_USEMMAP
+
+// On UNIX systems we use mmap and mprotect
+// On Windows this should be eventually extended to use
+// VirtualAlloc and VirtualProtect
+
+#include <sys/mman.h>
+#include <string.h>
+
+void *raw_data_copy(const void* data, HPy_ssize_t size, bool write_protect) {
+    void* new_ptr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    memcpy(new_ptr, data, size);
+    if (write_protect) {
+        mprotect(new_ptr, size, PROT_READ);
+    }
+    return new_ptr;
+}
+
```

Comment:
I think you addressed this concern in another comment but I wanted to double check. What happens if you `mmap()` lots of small buffers? The returned regions are aligned to page boundary (so, 4096 in general), but I assume that this consumes only the virtual address space and not the physical memory?

Reply:
> I think you cannot get smaller amount of physical memory than a page? 

indeed, I think you are correct.

> On the other hand, the number of the buffers is limited by the max number of handles, so it should be fine I believe [cut]

yes, this looks correct as well. I vote for including your computation/explanation in the "big docstring which explains how it works".

> This may mean that debug mode run may run out of memory when normal run wouldn't, but that may happen for other reasons too, so not sure if this is a big concern. The necessary amount of memory is "just" 2x, i.e., grows linearly, so I think it's fine?

yes, I think it's fine if debug mode requires more memory. Maybe at some point we can think of introducing optional checks (like, I want to enable leak checks but not raw buffers to save memory or so) but I don't think it's too important now.